File: src/main/osu/beatmap/parser/sections/BeatmapDifficultyParser.java
package main.osu.beatmap.parser.sections;

import main.osu.Utils;
import main.osu.beatmap.BeatmapData;

/**
 * A parser for parsing a beatmap's difficulty section.
 */
public class BeatmapDifficultyParser extends BeatmapKeyValueSectionParser {
    @Override
    public void parse(BeatmapData data, String line) {
        String[] p = splitProperty(line);

        switch (p[0]) {
            case "CircleSize":
                data.difficulty.cs = Utils.tryParseFloat(p[1], data.difficulty.cs);
                break;
            case "OverallDifficulty":
                data.difficulty.od = Utils.tryParseFloat(p[1], data.difficulty.od);
                if (Float.isNaN(data.difficulty.ar)) {
                    data.difficulty.ar = data.difficulty.od;
                }
                break;
            case "ApproachRate":
                data.difficulty.ar = Utils.tryParseFloat(p[1], Float.isNaN(data.difficulty.ar) ? data.difficulty.od : data.difficulty.ar);
                break;
            case "HPDrainRate":
                data.difficulty.hp = Utils.tryParseFloat(p[1], data.difficulty.hp);
                break;
            case "SliderMultiplier":
                data.difficulty.sliderMultiplier = Utils.tryParseFloat(p[1], data.difficulty.sliderMultiplier);
                break;
            case "SliderTickRate":
                data.difficulty.sliderTickRate = Utils.tryParseInt(p[1], data.difficulty.sliderTickRate);
        }
    }
}

File: src/com/rian/osu/beatmap/parser/sections/BeatmapDifficultyParser.kt
package com.rian.osu.beatmap.parser.sections

import com.rian.osu.beatmap.Beatmap
import kotlinx.coroutines.CoroutineScope

/**
 * A parser for parsing a beatmap's difficulty section.
 */
object BeatmapDifficultyParser : BeatmapKeyValueSectionParser() {
    override fun parse(beatmap: Beatmap, line: String, scope: CoroutineScope?) = splitProperty(line, scope)?.let {
        when (it.first) {
            "CircleSize" -> {
                val value = parseFloat(it.second)

                beatmap.difficulty.difficultyCS = value
                beatmap.difficulty.gameplayCS = value
            }

            "OverallDifficulty" -> beatmap.difficulty.od = parseFloat(it.second)
            "ApproachRate" -> beatmap.difficulty.ar = parseFloat(it.second)
            "HPDrainRate" -> beatmap.difficulty.hp = parseFloat(it.second)
            "SliderMultiplier" -> beatmap.difficulty.sliderMultiplier = parseDouble(it.second).coerceIn(0.4, 3.6)
            "SliderTickRate" -> beatmap.difficulty.sliderTickRate = parseDouble(it.second).coerceIn(0.5, 8.0)
        }
    } ?: throw UnsupportedOperationException("Malformed difficulty property: $line")
}
